### PR TITLE
[WIP] New type inference

### PR DESCRIPTION
**Pull request by gberaudo**
Switch to the new type inference version of the compiler.
See https://github.com/google/closure-compiler/wiki/Using-NTI-%28new-type-inference%29

There are only 2 errors left:
```
ERR! compile /home/gberaudo/dev/ol3/src/ol/observable.js:22: ERROR - ol.Observable cannot extend this type; structs can only extend structs
ERR! compile ol.Observable = function() {
ERR! compile                 ^
ERR! compile 
ERR! compile 
ERR! compile /home/gberaudo/dev/ol3/src/ol/renderer/maprenderer.js:38: ERROR - ol.renderer.Map cannot extend this type; structs can only extend structs
ERR! compile ol.renderer.Map = function(container, map) {
ERR! compile                   ^
ERR! compile 
ERR! compile 
ERR! compile 2 error(s), 0
ERR! compile  warning(s)
ERR! compile 
ERR! Process exited with non-zero status, see log for more detail: 2 
```

TODO:
- [x] fix `structs can only extend structs` error;
- [ ] fix publish.js

### CODE REVIEW

**Comment by gberaudo** [comment]
@fredj, if you want to have a look...

File: src/ol/events/event.js [hunk 1]
```diff
@@ -47,7 +47,7 @@ ol.events.Event.prototype.stopPropagation = function() {
 /**
  * @param {Event|ol.events.Event} evt Event
  */
-ol.events.Event.stopPropagation = function(evt) {
+ol.events.Event.prototype.stopPropagation = function(evt) {
```
Comment (tschaub) [inline src/ol/events/event.js]: This looks like a mistake to me.  There are still a number of places in the library that use the `ol.events.Event.stopPropagation` function.  With this change, that will be `undefined`.  I hope this causes at least a few test failures.

Not sure why you'd want it to be a prototype member.  A name like `ol.events.stopPropagation` could make sense, but it doesn't need to be accessible from `ol.events.Event` instances.

File: src/ol/events/event.js [hunk 2]
```diff
@@ -55,6 +55,6 @@ ol.events.Event.stopPropagation = function(evt) {
 /**
  * @param {Event|ol.events.Event} evt Event
  */
-ol.events.Event.preventDefault = function(evt) {
+ol.events.Event.prototype.preventDefault = function(evt) {
```
Comment (tschaub) [inline src/ol/events/event.js]: Same as above.  This is only used [one place](https://github.com/openlayers/ol3/blob/34d8d77344ee0b653770f065c593d4ab7b5d102b/src/ol/renderer/dom/dommaprenderer.js#L70-L72) in the library, but it cannot be renamed here without changing it there.  Again, if we are going to rename it, `ol.events.preventDefault` makes sense.  But again, it doesn't need to be accessible from every `ol.events.Event` instance.

File: src/ol/format/gml/gml3format.js [hunk 3]
```diff
@@ -1332,27 +1332,3 @@ ol.format.GML3.prototype.writeFeaturesNode = function(features, opt_options) {
...
-/**
- * Encode an array of features in GML 3.1.1 Simple Features.
- *
- * @function
- * @param {Array.<ol.Feature>} features Features.
- * @param {olx.format.WriteOptions=} opt_options Options.
- * @return {string} Result.
- * @api stable
- */
-ol.format.GML.prototype.writeFeatures;
```
Comment (tschaub) [inline src/ol/format/gml/gml3format.js]: Are we going to need another method of [documenting this method](http://openlayers.org/en/v3.14.1/apidoc/ol.format.GML3.html#writeFeatures)?

File: src/ol/format/gml/gml3format.js [hunk 4]
```diff
@@ -1332,27 +1332,3 @@ ol.format.GML3.prototype.writeFeaturesNode = function(features, opt_options) {
...
-/**
- * Encode an array of features in the GML 3.1.1 format as an XML node.
- *
- * @function
- * @param {Array.<ol.Feature>} features Features.
- * @param {olx.format.WriteOptions=} opt_options Options.
- * @return {Node} Node.
- * @api
- */
-ol.format.GML.prototype.writeFeaturesNode;
```
Comment (tschaub) [inline src/ol/format/gml/gml3format.js]: Same question here regarding [documentation](http://openlayers.org/en/v3.14.1/apidoc/ol.format.GML3.html#writeFeaturesNode).

**Comment by gberaudo** [comment]
The `structs can only extend structs` error seems to be gone after @tschaub's PR.
However there are lots of other errors, notably about nullable types.

**Comment by probins** [comment]
@gberaudo I notice a couple of typedef-related changes in here. First, LRUCache looks like it should be changed to an interface, so I won't include this in my typedef PR. Second, in `raster/operation.js` you're adding `goog.require('ol.raster.Pixel')` even though that is only used in comments/typedefs. Were you getting some sort of error here that prompted you to add the `goog.require()`?

**Comment by gberaudo** [comment]
@probins, thank you for caring about this PR.
I think it is better to ignore the changes here since some of them may be due to goog dependencies or the way the code is organized. Let's move the typedefs as discussed.

**Pull request closed without merging**